### ISSUE TITLE
training a language model is very slow

### ISSUE THREAD

**Issue by radekosmulski** [fastai/fastai#2812]
I tested this both locally (gcp instance with a single v100) and on colab (pip install from git repo of fastai).

The training of a LM is very slow - the slowness seems to lie in the construction of the batch.

![lm_training](https://user-images.githubusercontent.com/2444926/93466615-d0f66380-f8ec-11ea-8aae-544129ded444.png)

I created [a reproducer on colab](https://colab.research.google.com/drive/1UmK8B58R_9Ikuboevc0CsLg8lmLVZTli?usp=sharing). It includes data.

**Comment by radekosmulski** [comment]
I tried to troubleshoot this further. The example I tried from the docs (imdb) work okay. 

I was able to narrow down the issue to the quantity of the data. Batch generation time depends on the amount of total data. 

Below results of generating batches, first image with the entire wiki en dump. The 2nd example with 1/24th of the data.

The behavior I see in `htop` is a single core being maxed out.

![image](https://user-images.githubusercontent.com/2444926/93607160-3fa8ef00-f9c9-11ea-9e71-367a6d420a7c.png)

**Comment by morganmcg1** [comment]
@radekosmulski After a quick look it seems like its the parallel tokenization that really slows down for some reason. In your colab example for me the progress bar flies for the first half (up to step 24 of 48), however then it really seems to hit a brick wall.

This is where things slow down:

```
def parallel_tokenize(items, tok=None, rules=None, n_workers=defaults.cpus, **kwargs):
    "Calls optional `setup` on `tok` before launching `TokenizeWithRules` using `parallel_gen"
    if tok is None: tok = WordTokenizer()
    if hasattr(tok, 'setup'): tok.setup(items, rules)
    return parallel_gen(TokenizeWithRules, items, tok=tok, rules=rules, n_workers=n_workers, **kwargs)
```

I passed `tok=SubwordTokenizer()` to the TextBlock and that also was slow, but faster than the default WordTokenizer (Spacy)

I think like you said its just a function of data size and the creation of the vocab. I guess it should be faster on a second pass when everything is cached

**Comment by radekosmulski** [comment]
The problem is that training is extremely slow. If I understand correctly, at that point everything has been tokenized, but batch creation time unfortunately scales with the amount of data to the point where the model is untrainable.

**Comment by morganmcg1** [comment]
Ah understood now...hmmm...I wonder what is going on

**Comment by morganmcg1** [comment]
Just to put some numbers on this, the decrease in speed is really disproportionate to the increase in dataset size...

## Numbers

### Data sizes, train/val split (lines in .txt file)

- Mini: 75,055/155 

- Bigger Mini: 375,275 / 310

### Datablock creation

- Mini: 23min 15s

- Bigger Mini: 23min 

### 11 item dataloader loop

- Mini: 2.38s per loop

- Bigger Mini: 47.3s per loop

### Show Batch timings

- Mini: 0.042s

- Bigger Mini: 4.3s

### Train time (1 batch)

- Mini:  0h 8m

- Bigger Mini: 3h 15m (estimate)

**Comment by morganmcg1** [comment]
Just taking another peak at this, the slowness seems to come from the amount of data in the datasets, e.g. `dls.train_ds`

Here is the size of the items in `train_ds` with the IMDB example (`.show_batch(max_n=2)` takes **191 ms** on my machine):

<img width="194" alt="Screenshot 2020-11-10 at 13 56 40" src="https://user-images.githubusercontent.com/20516801/98726838-a361ef80-235c-11eb-8eba-8a0503bd1435.png">


And here is the size of the items in @radekosmulski 's `train_ds` (`.show_batch(max_n=2)` takes **19.5 s** on my machine):

<img width="310" alt="Screenshot 2020-11-10 at 13 58 43" src="https://user-images.githubusercontent.com/20516801/98727054-f3d94d00-235c-11eb-93cf-91087282b364.png">


(maybe this is obvious to folks, sharing just in case)

**Comment by jph00** [comment]
Yeah sorry I should have replied earlier - I see the problem. Numericalize is being called on the whole dataset each time. It should probably be a batch transform, or maybe part of the DataLoader. Or a preprocessing step.

**Comment by morganmcg1** [comment]
Tried a few things based on what you said:

### Moving Numericalize to `batch_tfms` in the init of TextBlock
Didn't work so well, had to modify a few functions that expected to receive a tensor but were now receiving tokenized text. Eventually hit a wall when it came to the `fa_collate` which again expects an array or tensor. When it didn't get either it turned a batch item into a list of tuples like so:

`[('xxbos',),('this',),('doesn't,),('work',)...]`

I think getting `batch_tfms` to work might end up being too hacky/fragile 

### Moving Numericalize to `item_tfms` in the init of TextBlock
Appears work work nicely, a quick test shows that timing is back down to **100-200ms** for `.show_batch`. I still have to test more thoroughly, including test that `reverse_text` works as expected

Had to modify 1 line of `create_item` in `LMDataLoader`:

```
def create_item(self:LMDataLoader, seq):
    #pdb.set_trace()
    if seq>=self.n: raise IndexError
    sl = self.last_len if seq//self.bs==self.n_batches-1 else self.seq_len
    st = (seq%self.bs)*self.bl + (seq//self.bs)*self.seq_len
    txt = self.chunks[st : st+sl+1]
    return (txt[:-1],txt[1:])     # <- New version
    #return LMTensorText(txt[:-1]),txt[1:]   # <- Old version
```

### Add to dataloader
Didn't try this yet, does a transform like `Numericalize` fit into the philosophy of what a Dataloader should do?

**Comment by jph00** [comment]
> Didn't try this yet, does a transform like `Numericalize` fit into the philosophy of what a Dataloader should do?


No not really - it wouldn't be a great approach.

**Comment by morganmcg1** [comment]
I spoke too soon, moving `Numericalize` to 'item_tfms' **does** speed things up when using a dataframe with large chunks of text in each row. 

But another issue with @radekosmulski 's example is that because its reading from files, and his files are quite huge.  `Tokenizer.encodes` reads every tokenized file and runs `.split(' ')`, like so:

```
    def encodes(self, o:Path):
        if self.mode=='folder' and str(o).startswith(str(self.path)):
            tok = self.output_dir/o.relative_to(self.path)
            return L(tok.read_text().split(' '))
        else: return self._tokenize1(o.read_text())
```

`snakeviz` profiler shows me it takes ~5s for .split and 2.3s for read_text

**Comment by morganmcg1** [comment]
# Potential Solutions

1. Use smaller text files :D

2. After a quick chat with Jeremy on the discord, a temporary work around would be do to the numericalization alongside the tokenization during `.setups`, and then keep these numericalized arrays in memory. Reading these in Tokenizer.encodes should then be pretty quick (hopefully)

Will give this a go now

**Comment by muellerzr** [comment]
@morganmcg1 did you ever wind up looking into this more? (I know, *really* old issue)
